src/modules/musig/main_impl.h

```diff
@@ -0,0 +1,566 @@
+        secp256k1_gej_add_ge_var(&combined_noncej, &combined_noncej, &noncep, NULL);
+    }
+    secp256k1_ge_set_gej(&combined_noncep, &combined_noncej);
+    if (secp256k1_fe_is_quad_var(&combined_noncep.y)) {
+        session->nonce_is_negated = 0;
+    } else {
+        session->nonce_is_negated = 1;
+        secp256k1_ge_neg(&combined_noncep, &combined_noncep);
+    }
+    if (nonce_is_negated != NULL) {
+        *nonce_is_negated = session->nonce_is_negated;
+    }
+    secp256k1_pubkey_save(&session->combined_nonce, &combined_noncep);
+    session->nonce_is_set = 1;
+    return 1;
+}
+
+int secp256k1_musig_session_set_msg(const secp256k1_context* ctx, secp256k1_musig_session *session, const unsigned char *msg32) {
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(session != NULL);
+    ARG_CHECK(msg32 != NULL);
+
+    if (session->msg_is_set) {
+        return 0;
+    }
+    memcpy(session->msg, msg32, 32);
+    session->msg_is_set = 1;
+    return 1;
+}
+
+int secp256k1_musig_partial_signature_serialize(const secp256k1_context* ctx, unsigned char *out32, const secp256k1_musig_partial_signature* sig) {
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(out32 != NULL);
+    ARG_CHECK(sig != NULL);
+    memcpy(out32, sig->data, 32);
+    return 1;
+}
+
+int secp256k1_musig_partial_signature_parse(const secp256k1_context* ctx, secp256k1_musig_partial_signature* sig, const unsigned char *in32) {
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(sig != NULL);
+    ARG_CHECK(in32 != NULL);
+    memcpy(sig->data, in32, 32);
+    return 1;
+}
+
+/* Compute msghash = SHA256(combined_nonce, combined_pk, msg) */
+static int secp256k1_musig_compute_messagehash(const secp256k1_context *ctx, unsigned char *msghash, const secp256k1_musig_session *session) {
+    unsigned char buf[33];
+    size_t bufsize = 33;
+    secp256k1_ge rp;
+    secp256k1_sha256 sha;
+
+    secp256k1_sha256_initialize(&sha);
+    secp256k1_pubkey_load(ctx, &rp, &session->combined_nonce);
+    secp256k1_fe_get_b32(buf, &rp.x);
+    secp256k1_sha256_write(&sha, buf, 32);
+    secp256k1_ec_pubkey_serialize(ctx, buf, &bufsize, &session->combined_pk, SECP256K1_EC_COMPRESSED);
+    VERIFY_CHECK(bufsize == 33);
+    secp256k1_sha256_write(&sha, buf, bufsize);
+    if (!session->msg_is_set) {
+        return 0;
+    }
+    secp256k1_sha256_write(&sha, session->msg, 32);
+    secp256k1_sha256_finalize(&sha, msghash);
+    return 1;
+}
+
+int secp256k1_musig_partial_sign(const secp256k1_context* ctx, const secp256k1_musig_session *session, secp256k1_musig_partial_signature *partial_sig) {
+    unsigned char msghash[32];
+    int overflow;
+    secp256k1_scalar sk;
+    secp256k1_scalar e, k;
+
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(partial_sig != NULL);
+    ARG_CHECK(session != NULL);
+
+    if (!session->nonce_is_set || !session->has_secret_data) {
+        return 0;
+    }
+
+    secp256k1_scalar_set_b32(&sk, session->seckey, &overflow);
+    if (overflow) {
+        return 0;
```

Comment:
We should probably zero out `sk` (and `k` below) in this and subsequent early-returns. Or even maintain a `ret` variable so there is only a single return.

Reply:
I didn't manage to get a readable implementation with a `ret` variable and didn't find a use of goto fail; in libsecp, but I added a commit that should best-effort-clear all secret scalars.

Reply:
Works for me.